### DISCUSSION TITLE
unable to add_chunk to existing file in Mojo::Asset::File

### DISCUSSION THREAD

**Discussion by s1037989** [mojolicious/mojo#1870] [Q&A] [answered]
Is it possible to add_chunk to an existing file with Mojo::Asset::File?

add_chunk to a non-existing file: **ok**
(notice the filename to use from the system command `mktemp`, but the file **is not** created)
```
$  perl -MMojo::Asset::File -E 'my $file = Mojo::Asset::File->new(path => $ARGV[0]); $file->add_chunk(rand); say $file->size;' `mktemp -u`
17
```

add_chunk to an existing file: **bad file descriptor**
(notice the filename to use is from the system command `mktemp`, and the filename **is** created)
```
$ perl -MMojo::Asset::File -E 'my $file = Mojo::Asset::File->new(path => $ARGV[0]); $file->add_chunk(rand); say $file->size;' `mktemp`
Can't write to asset: Bad file descriptor at -e line 1.
```

On that note, what would you think about adding add_chunk to Mojo::File?

add_chunk, similar to spurt, but open as append (>>):
```
sub add_chunk {
  my ($self, $content) = (shift, join '', @_);
  CORE::open my $file, '>>', $$self or croak qq{Can't open file "$$self": $!};
  ($file->syswrite($content) // -1) == length $content or croak qq{Can't write to file "$$self": $!};
  return $self;
}
```

**Comment by jhthorsen** [comment, marked as answer]
Your "add_chunk()" code can be written as `path($file)->open('>>')->syswrite($content)`.

You'll get that error message from Mojo::Asset::File when operating on existing files. I think Mojo::Asset::File is only designed to modify files if it's initialized with non-existing files. This makes sense when the webserver receives a huge body, where you the content is "new", but could be moved to a non-temporary location.

https://github.com/mojolicious/mojo/blob/main/lib/Mojo/Asset/File.pm#L15
